File: p7/p7.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// FIFO
int fifo(const vector<int>& pages, int capacity) {
    vector<int> frame;
    int faults = 0;

    for (int page : pages) {
        if (find(frame.begin(), frame.end(), page) == frame.end()) {
            if (frame.size() < capacity)
                frame.push_back(page);
            else {
                frame.erase(frame.begin()); // remove oldest
                frame.push_back(page);
            }
            faults++;
        }
        cout << "Page " << page << " -> ";
        for (int f : frame) cout << f << " ";
        cout << endl;
    }
    return faults;
}

// LRU
int lru(const vector<int>& pages, int capacity) {
    vector<int> frame, recent;
    int faults = 0;

    for (int page : pages) {
        auto it = find(frame.begin(), frame.end(), page);
        if (it == frame.end()) { // not found
            if (frame.size() < capacity)
                frame.push_back(page);
            else {
                int lru_page = recent.front();
                recent.erase(recent.begin());
                frame.erase(remove(frame.begin(), frame.end(), lru_page), frame.end());
                frame.push_back(page);
            }
            faults++;
        } else {
            // page already in frame → move it to most recent
            recent.erase(remove(recent.begin(), recent.end(), page), recent.end());
        }
        recent.push_back(page);

        cout << "Page " << page << " -> ";
        for (int f : frame) cout << f << " ";
        cout << endl;
    }
    return faults;
}

// Optimal
int optimal(const vector<int>& pages, int capacity) {
    vector<int> frame;
    int faults = 0;

    for (size_t i = 0; i < pages.size(); ++i) {
        int page = pages[i];
        if (find(frame.begin(), frame.end(), page) == frame.end()) {
            if (frame.size() < capacity)
                frame.push_back(page);
            else {
                int farthest = -1, replace_idx = -1;
                for (size_t j = 0; j < frame.size(); ++j) {
                    auto it = find(pages.begin() + i + 1, pages.end(), frame[j]);
                    int next_use = (it == pages.end()) ? pages.size() : distance(pages.begin(), it);
                    if (next_use > farthest) {
                        farthest = next_use;
                        replace_idx = j;
                    }
                }
                frame[replace_idx] = page;
            }
            faults++;
        }
        cout << "Page " << page << " -> ";
        for (int f : frame) cout << f << "";
        cout << endl;
    }
    return faults;
}

// Main
int main() {
    vector<int> pages = {7, 0, 1, 2, 0, 3, 0, 4, 2, 3, 0, 3, 2};
    int capacity = 3;

    cout << "Page Reference String: ";
    for (int p : pages) cout << p << " ";
    cout << "\nFrame Capacity: " << capacity << "\n\n";

    cout << "--- FIFO Page Replacement ---\n";
    cout << "Total Page Faults: " << fifo(pages, capacity) << "\n\n";

    cout << "--- LRU Page Replacement ---\n";
    cout << "Total Page Faults: " << lru(pages, capacity) << "\n\n";

    cout << "--- Optimal Page Replacement ---\n";
    cout << "Total Page Faults: " << optimal(pages, capacity) << "\n";

    return 0;
}

File: p7/p7.py
# Page Replacement Algorithm Simulation
# FIFO, LRU, and Optimal

def fifo(pages, capacity):
    frame = []
    page_faults = 0

    for page in pages:
        if page not in frame:
            if len(frame) < capacity:
                frame.append(page)
            else:
                frame.pop(0)
                frame.append(page)
            page_faults += 1
        print(f"Page {page} -> Frames: {frame}")
    return page_faults

def lru(pages, capacity):
    frame = []
    page_faults = 0
    recent = []

    for page in pages:
        if page not in frame:
            if len(frame) < capacity:
                frame.append(page)
            else:
                # Remove least recently used page
                lru_page = recent.pop(0)
                frame.remove(lru_page)
                frame.append(page)
            page_faults += 1
        else:
            recent.remove(page)
        recent.append(page)
        print(f"Page {page} -> Frames: {frame}")
    return page_faults

def optimal(pages, capacity):
    frame = []
    page_faults = 0

    for i in range(len(pages)):
        page = pages[i]
        if page not in frame:
            if len(frame) < capacity:
                frame.append(page)
            else:
                # Predict which page won’t be used for the longest time
                farthest_index = -1
                page_to_remove = None
                for f in frame:
                    if f not in pages[i+1:]:
                        page_to_remove = f
                        break
                    else:
                        next_use = pages[i+1:].index(f)
                        if next_use > farthest_index:
                            farthest_index = next_use
                            page_to_remove = f
                frame.remove(page_to_remove)
                frame.append(page)
            page_faults += 1
        print(f"Page {page} -> Frames: {frame}")
    return page_faults

def main():
    pages = [7, 0, 1, 2, 0, 3, 0, 4, 2, 3, 0, 3, 2]
    capacity = 3

    print("Page Reference String:", pages)
    print("Frame Capacity:", capacity)

    print("\n--- FIFO Page Replacement ---")
    faults = fifo(pages, capacity)
    print(f"Total Page Faults (FIFO): {faults}\n")

    print("--- LRU Page Replacement ---")
    faults = lru(pages, capacity)
    print(f"Total Page Faults (LRU): {faults}\n")

    print("--- Optimal Page Replacement ---")
    faults = optimal(pages, capacity)
    print(f"Total Page Faults (Optimal): {faults}\n")

if __name__ == "__main__":
    main()
